libraries/lib-utility/GlobalVariable.h

```diff
@@ -0,0 +1,167 @@
+      {
+         if constexpr (ScopedOnly)
+            Assign(std::move(m_previous));
+         else if (m_previous)
+            Assign(std::move(*m_previous));
+      }
+      /*! Won't compile unless ScopedOnly is false */
+      void Commit()
+      {
+         static_assert(!ScopedOnly);
+         m_previous.reset();
+      }
+      bool HasValue() const
+      {
+         if constexpr (ScopedOnly)
+            return true;
+         else
+            return m_previous.has_value();
+      }
+
+   private:
+      std::conditional_t<ScopedOnly, mutable_type, std::optional<mutable_type>>
+         m_previous;
+   };
+
+   /*! @brief Can guarantee that the global variable's lifetime encloses
+    those of other objects of static duration
+    
+    This is like the "nifty counter" idiom.  The guarantee is not automatic,
+    but you can define a `static` instance in a multiply included header file,
+    which creates multiple Initializers, but only one of them causes the
+    initial construction of the variable.
+
+    This makes it safe for other calls to Get() to occur inside the bodies
+    of both the constructor and destructor of another long-lived object.  Such
+    objects can be freely defined by other translation units that include the
+    header with no extra effort.
+    */
+   struct Initializer { Initializer() { Instance(); } };
+
+private:
+   //! Use static functions only.  Don't directly construct this.
+   GlobalVariable() = delete;
+
+   //! Generate the static variable
+#ifdef _WIN32
+   __declspec(dllexport)
+#endif
+   static mutable_type &Instance()
+   {
+      static_assert(!std::is_reference_v<stored_type>);
+      if constexpr (initializer != nullptr) {
+         static mutable_type instance{ initializer() };
+         return instance;
+      }
+      else {
+         static mutable_type instance;
+         return instance;
+      }
+   }
+
+   template<bool enable = true>
+   static auto Assign(mutable_type &&replacement)
+      -> std::enable_if_t<enable, mutable_type>
+   {
+      auto &instance = Instance();
+      auto result = std::move(instance);
+      instance = std::move(replacement);
+      return result;
+   }
+};
+
+//! Global function-valued variable, adding a convenient Call()
+template<typename Tag, typename Signature, auto... Options>
+class GlobalHook
+   : public GlobalVariable<Tag, const std::function<Signature>, Options...>
+{
+public:
+   using result_type = typename std::function<Signature>::result_type;
+
+   //! Null check of the installed function is done for you
+   /*! Requires that the return type of the function is void or
+    default-constructible */
+   template<typename... Arguments>
+   static result_type Call(Arguments &&...arguments)
+   {
+      auto &fn = GlobalHook::Get();
+      if (fn)
+         return fn(std::forward<Arguments>(arguments)...);
+      else if constexpr (std::is_void_v<result_type>)
+         return;
+      else
+         return result_type{};
```

Comment:
There is no way to know whether the function was actually invoked, in many cases `result_type{}` could be a valid value

Reply:
Maybe it makes sense to change the interface so that it becomes mandatory to initialize the hook with some meaningful value?

Reply:
This is all as I intend, to simplify the usage.  The hook can be called with `Call` without the complication of using `Get` and a null check first, provided that the function type returns void or a default-constructible type.  I think this convenience is good in most cases.

If in a particular case it is important to know whether any function was installed, the return type might be a pointer or a `std::optional`, while it is expected that an explicitly installed function will return non-null.  Then the caller can test the return value.

If there is a particular case in which it it is not good, then the template argument of `GlobalHook` can be a function returning a non-default-constructible type, and `Call` will fail to compile, compelling the more verbose usage instead.